### PR TITLE
Add support for patching the block ratelimiter

### LINKED ISSUE #1338: Patching drives write permissions
We currently have a use case where we need to dynamically attach drives, we do this by creating stub drives and later using patch to replace the stub drive with an actual drive. However, we've found that we want to give the option to our users to allow them to specify the rate limiter and write permissions during patching. We can work around this by doing it all in the beginning when we create the initial drives. However, this requires our users to know upfront what rate limiting and write permissions they need. While implementing this work around isn't too much work, having the ability to specify the rate limiters and write permissions during patch would really simplify things.

### PR DESCRIPTION
**Pull request by acatangiu**
## Reason for This PR

Partially fixes #1338 

## Description of Changes

Provided in commits.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] ~Any newly added `unsafe` code is properly documented.~
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: tests/integration_tests/functional/test_drives.py [hunk 1]
```diff
@@ -329,6 +329,107 @@ def test_patch_drive(test_microvm_with_ssh, network_config):
...
+
+    # "dd" writes to stderr by design. We drop first lines
+    stderr.readline().strip()
+    stderr.readline().strip()
+    dd_result = stderr.readline().strip()
+
+    # Interesting output looks like this:
+    # 4194304 bytes (4.2 MB, 4.0 MiB) copied, 0.0528524 s, 79.4 MB/s
+    tokens = dd_result.split()
+    print(tokens)
```
Comment (georgepisaltu) [inline tests/integration_tests/functional/test_drives.py:350]: Do we need this print?

File: tests/integration_tests/functional/test_api.py [hunk 2]
```diff
@@ -732,6 +732,57 @@ def _drive_patch(test_microvm):
     )
     assert test_microvm.api_session.is_status_no_content(response.status_code)
 
+    # Updates to valid `path_on_host` and `rate_limiter` are allowed.
```
Comment (georgepisaltu) [inline tests/integration_tests/functional/test_api.py:735]: nit: I would move the integration tests to a separate commit.

**Review by georgepisaltu — COMMENTED** [review]
LGTM! Just a couple of nits, and the coverage conflict.

File: src/api_server/src/request/drive.rs [hunk 3]
```diff
@@ -152,6 +157,48 @@ mod tests {
                 "path_on_host": "dummy"
               }"#;
         assert!(parse_patch_drive(&Body::new(body), Some(&"bar")).is_err());
+
```
Comment (alindima) [inline src/api_server/src/request/drive.rs:160]: would be nice to have a comment for each of these test cases, like we do for the other assertions above, to catch the essence of the test case

File: tests/integration_tests/functional/test_drives.py [hunk 4]
```diff
@@ -329,6 +329,107 @@ def test_patch_drive(test_microvm_with_ssh, network_config):
...
+                'size': 100,
+                'refill_time': 100
+            }
+        }
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+    test_microvm.start()
+    ssh_connection = net_tools.SSHConnection(test_microvm.ssh_config)
+
+    check_iops_limit(ssh_connection, 512, 1000, 0.85, 1.25)
```
Comment (alindima) [inline tests/integration_tests/functional/test_drives.py:395]: I might lack the context for rate-limiting, but I think that comments in regards to where these numbers come from would help

**Review by alindima — COMMENTED** [review]
LGTM, few nits about comments

File: tests/integration_tests/functional/test_drives.py (hunk 4 above)
Comment (sandreim) [reply tests/integration_tests/functional/test_drives.py:395]: I think it is clear if you look at the fn prototype and doc line:
```
def check_iops_limit(ssh_connection, block_size, count, min_time, max_time):
    """Verify if the rate limiter throttles block iops using dd."""
```

Based on the limiter confioguration we expect to do that much IO (count * block_size ) in a specific interval of time while allowing some variance.

File: tests/integration_tests/functional/test_drives.py (hunk 1 above)
Comment (sandreim) [reply tests/integration_tests/functional/test_drives.py:350]: I think this can be deleted.

**Review by sandreim — APPROVED** [review]
Looks good, aside for some nits that were pointed out already.

File: tests/integration_tests/functional/test_drives.py (hunk 1 above)
Comment (sandreim) [reply tests/integration_tests/functional/test_drives.py]: Deleted it.

File: src/api_server/src/request/drive.rs (hunk 3 above)
Comment (sandreim) [reply src/api_server/src/request/drive.rs:161]: Added comments.

File: tests/integration_tests/functional/test_drives.py (hunk 4 above)
Comment (sandreim) [reply tests/integration_tests/functional/test_drives.py:398]: Added comments.

**Comment by sandreim** [comment]
@alindima @georgepisaltu I've taken over this PR from @acatangiu and updated it. PTAL.

**Review by alindima — APPROVED** [review]

File: src/api_server/src/request/drive.rs [hunk 5]
```diff
@@ -112,28 +43,32 @@ pub fn parse_patch_drive(body: &Body, id_from_path: Option<&&str>) -> Result<Par
         return Err(Error::EmptyID);
     };
 
-    let patch_drive_payload = PatchDrivePayload {
-        fields: serde_json::from_slice(body.raw()).map_err(|e| {
-            METRICS.patch_api_requests.drive_fails.inc();
+    let block_device_update_cfg: BlockDeviceUpdateConfig =
+        serde_json::from_slice::<BlockDeviceUpdateConfig>(body.raw()).map_err(|e| {
+            METRICS.put_api_requests.drive_fails.inc();
```
Comment (georgepisaltu) [inline src/api_server/src/request/drive.rs:48]: Shouldn't this be `METRICS.patch_api_fails.drive_fails.in()`?

**Review by georgepisaltu — COMMENTED** [review]
Looks good, just one comment left.

File: src/api_server/src/request/drive.rs (hunk 5 above)
Comment (sandreim) [reply src/api_server/src/request/drive.rs]: Fixed.

**Review by acatangiu — COMMENTED** [review]
@georgepisaltu @alindima please take a look

**Review by alindima — APPROVED** [review]

**Review by georgepisaltu — APPROVED** [review]
